File: mcp_server_development/python/src/session14/probe_stdio.py
"""展開した配布物から Python 版サーバーを起動し、MCP で応答するかを確認する

  docker compose exec python python src/session14/probe_stdio.py \
      /tmp/verify/docsearch_mcp-0.1.0/src /app/src/session14/sample-docs

クライアント側のスクリプトなので print を使ってかまいません。
"""

from __future__ import annotations

import os
import sys
from pathlib import Path

import anyio
from mcp import ClientSession, StdioServerParameters
from mcp.client.stdio import stdio_client


async def main(package_src: Path, docs_root: Path) -> None:
    # 環境変数は自動で子プロセスに渡りません。ホストが渡すのと同じように、
    # 必要なものだけを明示します。これが「配布時の設定項目」の実体です。
    env = {
        "PATH": os.environ.get("PATH", "/usr/local/bin:/usr/bin:/bin"),
        "HOME": os.environ.get("HOME", "/root"),
        "PYTHONPATH": str(package_src),
        "DOCSEARCH_ROOT": str(docs_root),
    }
    params = StdioServerParameters(
        command=sys.executable,
        args=["-m", "docsearch_mcp"],
        env=env,
    )

    async with stdio_client(params) as (read, write):
        async with ClientSession(read, write) as session:
            init = await session.initialize()
            print(f"[probe] serverInfo: {init.server_info.name} v{init.server_info.version}")

            tools = (await session.list_tools()).tools
            print(f"[probe] tools: {', '.join(tool.name for tool in tools)}")

            result = await session.call_tool("search_documents", {"query": "VPN"})
            structured = result.structured_content or {}
            hits = structured.get("results") or []
            first = hits[0]["path"] if hits else None
            print(
                f'[probe] search_documents("VPN"): '
                f'totalMatched={structured.get("totalMatched")} / 先頭={first}'
            )

    print("OK: 展開した配布物から起動したサーバーが MCP として応答しました")


if __name__ == "__main__":
    if len(sys.argv) < 3:
        print(
            "使い方: python src/session14/probe_stdio.py <展開先の src> <検索対象ディレクトリ>",
            file=sys.stderr,
        )
        raise SystemExit(2)
    anyio.run(main, Path(sys.argv[1]).resolve(), Path(sys.argv[2]).resolve())

File: mcp_server_development/node/src/session14/probe-stdio.ts
/**
 * 配布物から起動したサーバーに MCP クライアントとして話しかける
 *
 *   docker compose exec node npx tsx src/session14/probe-stdio.ts docsearch-mcp --docs-root /app/src/session14/sample-docs
 *   docker compose exec -e DOCSEARCH_ROOT=/app/src/session14/sample-docs node npx tsx src/session14/probe-stdio.ts docsearch-mcp
 *
 * 第 1 引数がコマンド、以降が引数です。ホストが渡すのと同じ command / args / env の形に
 * そろえてあるので、「ホスト設定ファイルに書く内容」をそのまま試せます。
 */
import { Client } from "@modelcontextprotocol/sdk/client/index.js";
import { StdioClientTransport } from "@modelcontextprotocol/sdk/client/stdio.js";

const [command, ...args] = process.argv.slice(2);
if (command === undefined) {
  console.error("使い方: npx tsx src/session14/probe-stdio.ts <コマンド> [引数...]");
  process.exit(2);
}

/**
 * 環境変数は子プロセスへ自動では渡りません。SDK は安全な一部だけを引き継ぐので、
 * サーバーに渡したい変数は明示します（ホスト設定の env がこれに相当します）。
 */
const docsRoot = process.env["DOCSEARCH_ROOT"];
const transport = new StdioClientTransport({
  command,
  args,
  ...(docsRoot === undefined
    ? {}
    : {
        env: {
          PATH: process.env["PATH"] ?? "",
          HOME: process.env["HOME"] ?? "",
          DOCSEARCH_ROOT: docsRoot,
        },
      }),
});

const client = new Client({ name: "session14-probe", version: "1.0.0" });
await client.connect(transport);

console.log(`[probe] 起動コマンド: ${[command, ...args].join(" ")}`);
const info = client.getServerVersion();
console.log(`[probe] serverInfo: ${info?.name} v${info?.version}`);

const { tools } = await client.listTools();
console.log(`[probe] tools: ${tools.map((tool) => tool.name).join(", ")}`);

const result = await client.callTool({
  name: "search_documents",
  arguments: { query: "VPN" },
});
const structured = result.structuredContent as
  | { totalMatched?: number; results?: Array<{ path: string }> }
  | undefined;
console.log(
  `[probe] search_documents("VPN"): totalMatched=${structured?.totalMatched}` +
    ` / 先頭=${structured?.results?.[0]?.path}`,
);

await client.close();
console.log("OK: 配布物から起動したサーバーが MCP として応答しました");
